### DISCUSSION TITLE
Is it possible to retrieve Client information in case of multiple clients at the server side?

### DISCUSSION THREAD

**Discussion by vk-gst** [Kistler-Group/sdbus-cpp#187] [Q&A] [answered]
I have an application that contains multiple clients(proxies) interacting with server application. They have a common object path for now.  

I wanted to know if it's possible to retrieve the client or proxy instance information(for eg. client-ids) at the server side, so as to associate the requests made by client1, client2, etc.  
One way, I could think of is to have unique object paths for each client instance and use this at the server side. But is there any other way with sdbus-c++ that provides me this information?

**Comment by riuriuriu** [comment]
As far as I know, D-Bus only knows a [connection's unique name](https://dbus.freedesktop.org/doc/dbus-specification.html#message-bus-names). But that's not available at the server/adapter side in sdbus-cpp (it provides a high level API). 

Is it just for debugging/monitoring? E.g. to sort out misbehaving clients? Then using a bus monitor tool like `dbus-monitor` would fit better.

**Comment by sangelovic** [comment, marked as answer]
@vk-gst Yes, a D-Bus message carries a `sender` field, which you can query through `Message::getSender()`. According to [D-Bus specification](https://dbus.freedesktop.org/doc/dbus-specification.html), this is a unique name of the sending connection. So with this, you'd able to differentiate clients invoking a server method.

**Reply by sangelovic** [reply]
@vk-gst So you simply invoke `getSender()` on a `MethodCall` message you get in your server method callback. This message is however only available in basic API level of sdbus-c++. If you use convenience API or generated stub-based API, you don't have access to the underlying message. But this is probably going to be changed soon, I think I will prepare a PR that will make this possible.
